### PR TITLE
Added create and free device driver registry

### PR DESCRIPTION
**Pull request by ewconnell**
Added functions to allocate and initialize a device driver registry, and to free it.

This is my first PR, should I be added as a Collaborator?
Oh damn, so it looks like separate pull requests need to be on different branches otherwise they get bundled together?

I fixed a separate issue of an incorrect buffer size calculation

### CODE REVIEW

**Comment by ScottTodd** [comment]
> This is my first PR, should I be added as a Collaborator?

Added! You should be able to request review, add labels, etc. now. Kokoro presubmit checks should also run automatically now (might need a poke this time though).

**Comment by ScottTodd** [comment]
> Oh damn, so it looks like separate pull requests need to be on different branches otherwise they get bundled together?

Yep, all commits on a branch between the base and the tip of the branch will be included in a PR. That's a feature you can use if you want (grouping several commits into one review). If the commits are independent then you can split them across several branches and send distinct PRs. You can stack changes onto several PRs pointed at each other if you use branches on the main repository, but that creates some extra noise on the main repo.

**Comment by ewconnell** [comment]
Thanks, I'll be careful to split them onto separate branches next time.

**Comment by ewconnell** [comment]
Hey, there is a check failure from clang_format but I don't understand the output. It doesn't seem to explain the error, it looks like clang_format just bombed and exited the process. Ideas?? I'm not sure how to track this down.

**Comment by ScottTodd** [comment]
> Hey, there is a check failure from clang_format but I don't understand the output. It doesn't seem to explain the error, it looks like clang_format just bombed and exited the process. Ideas?? I'm not sure how to track this down.

https://github.com/google/iree/pull/5116/checks?check_run_id=2116950783 these logs look actionable? They show the diff from `git-clang-format`.

File: iree/base/api.c [hunk 1]
```diff
@@ -636,20 +636,18 @@ iree_status_allocate(iree_status_code_t code, const char* file, uint32_t line,
...
   va_end(varargs_0);
-  va_end(varargs_1);
   return ret;
 }
 
-IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
-iree_status_allocate_vf(iree_status_code_t code, const char* file,
-                        uint32_t line, const char* format, va_list varargs_0,
-                        va_list varargs_1) {
+iree_status_t _iree_status_allocate_vf(iree_status_code_t code,
```
Comment (benvanik) [inline iree/base/api.c:647]: make functions that are private to a compilation unit (.c) `static`, like `static iree_status_t ...` - this ensures the function has internal linkage and can't be referenced from any other file (even if the name collides)

File: iree/base/api.c [hunk 2]
```diff
@@ -854,6 +863,20 @@ iree_status_annotate_vf(iree_status_t base_status, const char* format,
 #endif  // has IREE_STATUS_FEATURE_ANNOTATIONS
 }
 
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_annotate_vf(iree_status_t base_status, const char* format,
+                        va_list varargs_0) {
+  // Issue #5184
+  // make a copy so we can walk the arguments once to determine storage size
+  // and once to copy the values.
```
Comment (benvanik) [inline iree/base/api.c:871]: remove the issue # here and above - the comment is enough

File: iree/base/api.c [hunk 3]
```diff
@@ -953,7 +976,9 @@ iree_status_to_string(iree_status_t status, char** out_buffer,
                                         /*buffer=*/NULL, &buffer_length))) {
     return false;
   }
-  char* buffer = (char*)malloc(buffer_length + 1);
+  // add space for NULL and pass this information on to iree_status_format
```
Comment (benvanik) [inline iree/base/api.c:979]: proper punctuation on comments here and elsewhere (`// Add a space for .... to iree_status_format.`) https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

File: iree/base/api.h [hunk 4]
```diff
@@ -724,17 +723,22 @@ IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
 
 IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
 iree_status_annotate_vf(iree_status_t base_status, const char* format,
-                        va_list varargs_0, va_list varargs_1);
+                        va_list varargs_0);
```
Comment (benvanik) [inline iree/base/api.h:726]: you can rename this to just varargs now (_0 and _1 were disambiguators for when there were two)

File: iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp [hunk 5]
```diff
@@ -81,28 +81,37 @@ class UnixLinkerTool : public LinkerTool {
...
-        // lowering paths introduce a bunch, though, so this is what we are
-        // stuck with.
-        "-shared",
-        "-undefined suppress",
-
         "-o " + artifacts.libraryFile.path,
     };
 
     if (targetTriple.isOSDarwin() || targetTriple.isiOS()) {
+      // Statically link all dependencies so we don't have any runtime deps.
```
Comment (benvanik) [inline iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp:88]: these seem unrelated? possibly due to merge issues?

File: iree/hal/driver_registry.h [hunk 6]
```diff
@@ -102,6 +102,17 @@ typedef struct iree_hal_driver_registry_s iree_hal_driver_registry_t;
 IREE_API_EXPORT iree_hal_driver_registry_t* IREE_API_CALL
 iree_hal_driver_registry_default();
 
+// Creates and initializes an empty device driver registry
+//
+// Thread-safe.
+IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_driver_registry_create(
```
Comment (benvanik) [inline iree/hal/driver_registry.h:108]: `_create` is reserved for reference counted objects (we roughly follow the core foundation naming and rules, so the names are meaningful: https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html)

Since this is just a non-refcounted instance this should be `_allocate` and the free should just be `_free` (no need for `unregister_and_`).

File: iree/hal/driver_registry.h [hunk 7]
```diff
@@ -102,6 +102,17 @@ typedef struct iree_hal_driver_registry_s iree_hal_driver_registry_t;
...
+    iree_allocator_t allocator, iree_hal_driver_registry_t** out_registry);
+
+// Unregisters each factory and frees the associated registry resources.
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_hal_driver_registry_unregister_and_free(
```
Comment (benvanik) [inline iree/hal/driver_registry.h:113]: the allocator passed to _create should be stored on the registry such that this does not need it to be passed in

File: iree/hal/driver_registry.c [hunk 8]
```diff
@@ -60,6 +60,39 @@ iree_hal_driver_registry_default() {
   return &iree_hal_driver_registry_default_;
 }
 
+// Creates and initializes an empty device driver registry
+IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_driver_registry_create(
+    iree_allocator_t allocator, iree_hal_driver_registry_t** out_registry) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(
+      allocator, sizeof(struct iree_hal_driver_registry_s),
```
Comment (benvanik) [inline iree/hal/driver_registry.c:69]: `sizeof(iree_hal_driver_registry_t)`

File: iree/hal/driver_registry.c [hunk 9]
```diff
@@ -60,6 +60,39 @@ iree_hal_driver_registry_default() {
   return &iree_hal_driver_registry_default_;
 }
 
+// Creates and initializes an empty device driver registry
+IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_driver_registry_create(
+    iree_allocator_t allocator, iree_hal_driver_registry_t** out_registry) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(
```
Comment (benvanik) [inline iree/hal/driver_registry.c:68]: `IREE_END_ZONE_AND_RETURN_IF_ERROR` (otherwise the zone above won't be ended if the alloc fails)

File: iree/hal/driver_registry.c [hunk 10]
```diff
@@ -60,6 +60,39 @@ iree_hal_driver_registry_default() {
...
+
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_hal_driver_registry_unregister_and_free(
+    iree_hal_driver_registry_t* registry, iree_allocator_t allocator) {
+  IREE_ASSERT_ARGUMENT(registry);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // unregister each factory
+  for (iree_host_size_t i = 0; i < registry->factory_count; ++i) {
+    IREE_RETURN_IF_ERROR(iree_hal_driver_registry_unregister_factory(
```
Comment (benvanik) [inline iree/hal/driver_registry.c:85]: `IREE_IGNORE_ERROR` and make this function return `void` - cleanup functions should not fail
(unregister can only fail if the factory provided is not registered, which here won't be the case as you are passing it in right from the list)

File: iree/hal/driver_registry.c [hunk 11]
```diff
@@ -60,6 +60,39 @@ iree_hal_driver_registry_default() {
...
+}
+
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_hal_driver_registry_unregister_and_free(
+    iree_hal_driver_registry_t* registry, iree_allocator_t allocator) {
+  IREE_ASSERT_ARGUMENT(registry);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // unregister each factory
+  for (iree_host_size_t i = 0; i < registry->factory_count; ++i) {
```
Comment (benvanik) [inline iree/hal/driver_registry.c:84]: this loop won't work as-is because the unregister mutates the factory_count and the list - you will need to do this backwards (count -> 0) so you are unregistering the tail each iteration or change it to a while-loop while the factory_count > 0

**Review by benvanik — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp (hunk 5 above)
Comment (ewconnell) [reply iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp:88]: Yes sorry, I thought a PR code review would only take a day or two. Scott wanted me to merge in his changes so that I could successfully build, and this ended up piggy backing. I wasn't sure how to isolate the first set of changes after I had already submitted the PR. I will be aware of this possibility in the future.

Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp:88]: A few options:

A) Merge `google:main` into your PR branch (which ideally wouldn't also be `main`, but too late to change that now :P). This will preserve all history.
B) Rebase your PR branch onto `google:main`, editing history (e.g. dropping `astrotuna proposed MacOS linker problem fix` and `update linker settings for macOS`, or keeping those commits and resolving the merge conflicts with the similar changes made to `google:main`)

Ultimately, you should either fixup the history on your PR _or_ merge with "squash and merge". Both are fine.

If there are actually more changes you need made to `UnixLinkerTool.cpp` for Apple platforms, those should be made on a separate PR.

**Comment by ScottTodd** [comment]
Is this ready for another look? I see a "changes to meet PR edit requests" commit :)

you can re-request review to bump it back to the "awaiting review from you" page at https://github.com/google/iree/pulls/review-requested/@me and send an email to reviewers:

image

**Comment by ewconnell** [comment]
The pre-submit checks are failing related to swiftshader, but it doesn't appear to have anything to do with my changes.
Is it possible for people to check things in that don't pass? I'm wondering how this happened.

**Comment by ScottTodd** [comment]
> The pre-submit checks are failing related to swiftshader, but it doesn't appear to have anything to do with my changes.
> Is it possible for people to check things in that don't pass? I'm wondering how this happened.

Yes. Our presubmit checks are not blocking and sometimes miss things (they run on the PR branch, which may fall behind mainline). The test failures here should be fixed already on https://github.com/google/iree/pull/5255. You can ignore them. The builds are named "swiftshader" since they use SwiftShader's software Vulkan driver instead of native GPU drivers, but they are not testing SwiftShader or Vulkan tests exclusively.

File: iree/base/api.h (hunk 4 above)
Comment (ScottTodd) [reply iree/base/api.h:726]: +1 (here and above)

File: iree/base/api.c [hunk 12]
```diff
@@ -696,6 +695,19 @@ iree_status_allocate_vf(iree_status_code_t code, const char* file,
 #endif  // has IREE_STATUS_FEATURE_ANNOTATIONS
 }
 
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_allocate_vf(iree_status_code_t code, const char* file,
+                        uint32_t line, const char* format, va_list varargs) {
+  // make a copy so we can walk the arguments once to determine storage size
+  // and once to copy the values.
```
Comment (ScottTodd) [inline iree/base/api.c:702]:
```suggestion
  // Make a copy so we can walk the arguments once to determine storage size
  // and once to copy the values.
```

style nit: capitalize the first word in the comment

File: iree/base/api.c [hunk 13]
```diff
@@ -854,6 +864,19 @@ iree_status_annotate_vf(iree_status_t base_status, const char* format,
 #endif  // has IREE_STATUS_FEATURE_ANNOTATIONS
 }
 
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_annotate_vf(iree_status_t base_status, const char* format,
+                        va_list varargs) {
+  // make a copy so we can walk the arguments once to determine storage size
+  // and once to copy the values.
```
Comment (ScottTodd) [inline iree/base/api.c:871]:
```suggestion
  // Make a copy so we can walk the arguments once to determine storage size
  // and once to copy the values.
```

same style nit

File: iree/hal/driver_registry.h [hunk 14]
```diff
@@ -102,6 +102,16 @@ typedef struct iree_hal_driver_registry_s iree_hal_driver_registry_t;
 IREE_API_EXPORT iree_hal_driver_registry_t* IREE_API_CALL
 iree_hal_driver_registry_default();
 
+// Creates and initializes an empty device driver registry
```
Comment (ScottTodd) [inline iree/hal/driver_registry.h:105]:
```suggestion
// Creates and initializes an empty device driver registry.
```

style nit: add period

(higher bar for comment style in API header files, sorry D:)

**Review by ScottTodd — APPROVED** [review]
LGTM with comments addressed (though I would wait for Ben's stamp also)

**Comment by benvanik** [comment]
This was implemented in #7974.
